iphone/cli/hooks/frameworks.js

```diff
@@ -0,0 +1,841 @@
+		this._metadataPathAndFilename = null;
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	get incrementalOutputs() {
+		return [ this._outputDirectory ];
+	}
+
+	/**
+	 * Sets the output directory where this task will write the framework metadata
+	 *
+	 * @param {String} outputPath Full path to the output directory
+	 */
+	set outputDirectory(outputPath) {
+		this._outputDirectory = outputPath;
+		this.registerOutputPath(outputPath);
+		this._metadataPathAndFilename = path.join(this._outputDirectory, 'frameworks.json');
+	}
+
+	/**
+	 * Returns a list with metadata of all recognized frameworks
+	 *
+	 * @return {Map.<String, FrameworkInfo>} Map of framework paths and the associated metadata
+	 */
+	get frameworks() {
+		return this._frameworks;
+	}
+
+	/**
+	 * Adds a .framework folder so this task can inspect it to collect metadata
+	 * about the framework
+	 *
+	 * @param {String} frameworkPath Path to the .framework folder to inspect
+	 */
+	addFrameworkPath(frameworkPath) {
+		if (this._frameworkPaths.has(frameworkPath)) {
+			return;
+		}
+
+		this._frameworkPaths.add(frameworkPath);
+		this.addInputDirectory(frameworkPath);
+	}
+
+	/**
+	 * Does a full task run by inspecting all available framework paths
+	 *
+	 * @return {Promise}
+	 */
+	doFullTaskRun() {
+		this._frameworks = new Map();
+		return this.inspectFrameworks(this._frameworkPaths)
+			.then(this.writeFrameworkMetadata.bind(this));
+	}
+
+	/**
+	 * Does an incremental task run by only scanning changed framework folders
+	 * and removing deleted frameworks from the metadata object
+	 *
+	 * @param {Map.<String, String>} changedFiles Map of changed files and their status (created, changed or deleted)
+	 * @return {Promise}
+	 */
+	doIncrementalTaskRun(changedFiles) {
+		let loaded = this.loadFrameworkMetadata();
+		if (!loaded) {
+			return this.doFullTaskRun();
+		}
+
+		this._frameworks.forEach(frameworkInfo => {
+			if (!fs.existsSync(frameworkInfo.path)) {
+				this.logger.trace(`Framework at ${frameworkInfo.path} deleted, removing metadata`);
+				this._frameworks.delete(frameworkInfo.name);
+				return;
+			}
+
+			// Remove any frameworks from deactivated modules
+			if (!this._frameworkPaths.has(frameworkInfo.path)) {
+				this.logger.trace(`Framework at ${frameworkInfo.path} no longer in search path, removing metadata`);
+				this._frameworks.delete(frameworkInfo.name);
+				return;
+			}
+		});
+
+		let changedFrameworks = new Set();
+		changedFiles.forEach((fileStatus, pathAndFilename) => {
+			if (fileStatus === 'created' || fileStatus === 'changed') {
+				let frameworkPath = pathAndFilename.substring(0, pathAndFilename.indexOf('.framework') + 10);
+				if (!changedFrameworks.has(frameworkPath)) {
+					this.logger.trace(`Framework at ${frameworkPath} changed, regenerating metadata`);
+					changedFrameworks.add(frameworkPath);
+				}
+			}
+		});
+
+		return this.inspectFrameworks(changedFrameworks)
+			.then(this.writeFrameworkMetadata.bind(this));
```

Comment:
Tip: instead of doing a `bind()`, you can just do:

```
.then(() => this.writeFrameworkMetadata());
```

Reply:
Nice, i've been looking for an elegant solution to get rid of `.bind` and this looks much better!